In [1]:
import pandas as pd
import networkx as nx

In [2]:
df = pd.read_csv('networks-id-volume1.csv')

G = nx.Graph()

for _, row in df.iterrows():
    G.add_edge(row['IdSource'], row['IdTarget'], weight=row['Weight'])

degree_centrality = nx.degree_centrality(G)
betweenness_centrality = nx.betweenness_centrality(G, weight='weight')
closeness_centrality = nx.closeness_centrality(G)
eigenvector_centrality = nx.eigenvector_centrality(G, weight='weight')

centrality_df = pd.DataFrame({
    'Character': list(degree_centrality.keys()),
    'Degree': list(degree_centrality.values()),
    'Betweenness': [betweenness_centrality[n] for n in degree_centrality],
    'Closeness': [closeness_centrality[n] for n in degree_centrality],
    'Eigenvector': [eigenvector_centrality[n] for n in degree_centrality]
})


centrality_df = centrality_df.sort_values(by='Eigenvector', ascending=False)

In [3]:
centrality_df

Character    Degree  Betweenness  Closeness  Eigenvector
0       frod  0.901639     0.005646   0.910448     0.564258
1       sams  0.606557     0.002704   0.709302     0.387704
2      ganda  0.885246     0.011520   0.897059     0.327409
3       arag  0.786885     0.003225   0.824324     0.299494
5       pipp  0.459016     0.006754   0.642105     0.248857
..       ...       ...          ...        ...          ...
61      sfax  0.114754     0.015331   0.530435     0.003176
54     treeb  0.131148     0.002099   0.516949     0.002981
58      thra  0.131148     0.003835   0.516949     0.002311
57      osgi  0.245902     0.023884   0.554545     0.001218
60      edor  0.016393     0.000000   0.386076     0.000027

[62 rows x 5 columns]

frod (Frodo) tem a maior centralidade de grau (0.90), indicando que ele é o personagem mais diretamente conectado com outros. ganda (Gandalf), o anel (ring), arag (Aragorn) e os elfos também estão entre os mais conectados. Frodo está no centro da trama, interagindo com quase todos. Isso faz sentido narrativamente, já que ele é o protagonista.

O anel (ring) (0.0539) tem uma das maiores intermediação — faz sentido se ele conecta grupos que não se relacionariam diretamente, já que o Anel é um elo entre muitos personagens. Elfos e sfax também têm valores altos, indicando que são intermediários entre diferentes grupos. O Anel conecta muitas tramas e personagens, mesmo que indiretamente. Alguns personagens com pouca conexão direta ainda são importantes para conectar partes da rede.

O frod (0.9104) e ganda (0.8970) estão no topo na medida de centralidade de proximidade (Closeness) — estão próximos de todos na rede, conseguindo alcançar qualquer outro personagem com poucos passos. Mostra que Frodo e Gandalf estão bem posicionados para transmitir ou acessar informações rapidamente na rede.

O frod (0.564) e sams (0.387) estão no topo da centralidade de autovetor. Frodo está não só conectado com muitos, mas com os personagens mais importantes da rede. Gandalf também tem grande influência. Já ring tem um valor menor (0.192), o que faz sentido se ele conecta muitos, mas não os mais influentes diretamente.